### PR TITLE
Fix flaky test failure in ServerCommandTest

### PR DESCRIPTION
**Pull request by ajaybhargavb**
#2303

### CODE REVIEW

File: test/Microsoft.AspNetCore.Razor.Tools.Test/ServerCommandTest.cs [hunk 1]
```diff
@@ -13,23 +13,23 @@ namespace Microsoft.AspNetCore.Razor.Tools
...
+            var path = Path.Combine(directoryPath, expectedFileName);
 
             var pipeName = Guid.NewGuid().ToString();
             var server = GetServerCommand(pipeName);
 
             // Act & Assert
             try
             {
-                using (var _ = server.WritePidFile())
+                using (var _ = server.WritePidFile(directoryPath))
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Razor.Tools.Test/ServerCommandTest.cs]: @pranavkm, I didn't need the Func<string, string> because we are passing the path directly.

File: src/Microsoft.AspNetCore.Razor.Tools/ServerCommand.cs [hunk 2]
```diff
@@ -119,20 +119,8 @@ protected virtual void ExecuteServerCore(ConnectionHost host, CompilerHost compi
             dispatcher.Run();
         }
 
-        internal FileStream WritePidFile()
+        internal virtual FileStream WritePidFile()
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Razor.Tools/ServerCommand.cs]: `protected virtual`?

File: test/Microsoft.AspNetCore.Razor.Tools.Test/ServerCommandTest.cs (hunk 1 above)
Comment (pranavkm) [reply test/Microsoft.AspNetCore.Razor.Tools.Test/ServerCommandTest.cs]: That was to verify the path calculation. Would be good add a test to cover that too

File: test/Microsoft.AspNetCore.Razor.Tools.Test/ServerCommandTest.cs [hunk 3]
```diff
@@ -47,10 +47,10 @@ public void WritePidFile_WorksAsExpected()
             }
             finally
             {
-                // Delete the file in case the test fails.
-                if (File.Exists(path))
+                // Cleanup after the test.
+                if (Directory.Exists(directoryPath))
                 {
-                    File.Delete(path);
+                    Directory.Delete(directoryPath);
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Razor.Tools.Test/ServerCommandTest.cs]: Might have to say `recursive: true`. I don't think `Directory.Delete` deletes files by default

File: test/Microsoft.AspNetCore.Razor.Tools.Test/ServerCommandTest.cs (hunk 1 above)
Comment (ajaybhargavb) [reply test/Microsoft.AspNetCore.Razor.Tools.Test/ServerCommandTest.cs]: Added.

File: src/Microsoft.AspNetCore.Razor.Tools/ServerCommand.cs [hunk 4]
```diff
@@ -133,7 +149,26 @@ internal FileStream WritePidFile()
...
+                writer.Write(content);
+            }
+
+            return fileStream;
+        }
+
+        // Internal for testing.
+        internal virtual string GetPidFilePath(Func<string, string> getEnvironmentVariable)
+        {
+            var path = getEnvironmentVariable?.Invoke("DOTNET_BUILD_PIDFILE_DIRECTORY");
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Razor.Tools/ServerCommand.cs]: Why would `getEnvironmentVariable` be null?

File: src/Microsoft.AspNetCore.Razor.Tools/ServerCommand.cs [hunk 5]
```diff
@@ -133,7 +149,26 @@ internal FileStream WritePidFile()
...
             if (string.IsNullOrEmpty(path))
             {
                 var homeEnvVariable = PlatformInformation.IsWindows ? "USERPROFILE" : "HOME";
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Razor.Tools/ServerCommand.cs]: `var homePath = getEnvironmentVariable(homeEnvVariable);`

File: test/Microsoft.AspNetCore.Razor.Tools.Test/ServerCommandTest.cs [hunk 6]
```diff
@@ -47,15 +47,44 @@ public void WritePidFile_WorksAsExpected()
...
             }
         }
 
-        private ServerCommand GetServerCommand(string pipeName)
+        [Fact]
+        public void GetPidFilePath_ReturnsCorrectDefaultPath()
+        {
+            // Arrange
+            var homeEnvVariable = PlatformInformation.IsWindows ? "USERPROFILE" : "HOME";
+            var expectedPath = Path.Combine(Environment.GetEnvironmentVariable(homeEnvVariable), ".dotnet", "pids", "build");
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Razor.Tools.Test/ServerCommandTest.cs]: Use a test function rather than do this especially since you control the result of `getEnv` now

File: test/Microsoft.AspNetCore.Razor.Tools.Test/ServerCommandTest.cs [hunk 7]
```diff
@@ -47,15 +47,44 @@ public void WritePidFile_WorksAsExpected()
...
+
+            // Act
+            var directoryPath = server.GetPidFilePath(getEnvironmentVariable: null);
+
+            // Assert
+            Assert.Equal(expectedPath, directoryPath);
+        }
+
+        [Fact]
+        public void GetPidFilePath_UsesEnvironmentVariablePathIfSpecified()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Razor.Tools.Test/ServerCommandTest.cs]: Also add a test for when `getEnvVariable` returns null.

File: src/Microsoft.AspNetCore.Razor.Tools/ServerCommand.cs (hunk 4 above)
Comment (ajaybhargavb) [reply src/Microsoft.AspNetCore.Razor.Tools/ServerCommand.cs]: It won't. Just being defensive. Better than letting it throw.

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Razor.Tools/ServerCommand.cs]: Nah

**Comment by ajaybhargavb** [comment]
🆙 📅
